### PR TITLE
Week 5: Test-Driven Development Meeting Scheduling Exercise

### PR DESCRIPTION
**Pull request by chrismohrii**
Adds a solution to the query method, more tests for mandatory attendees, and tests for optional attendees.

### CODE REVIEW

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 1]
```diff
@@ -28,6 +28,16 @@ public void testGreeting() {
 
     String greeting = greeter.greet("Ada");
 
+    Assert.assertEquals("Hello Ada", greeting);
+  }
+
+	@Test
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: nit: Let's fix the alignment for this `@Test` annotation.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,101 @@
...
+      return queryEveryone;
+    }
+    else {
+      return queryGivenAttendees(events, request);
+    }
+  }
+
+  /** Returns the possible meeting times given required attendees */
+  public Collection<TimeRange> queryGivenAttendees(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> busy = new ArrayList<>();
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: Consider more descriptive variable names like `busyTimes` or `unavailableTimes`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,101 @@
...
+    }
+
+    // Sort these times by start time
+    Collections.sort(busy, TimeRange.ORDER_BY_START);
+
+    // Simplify the times to avoid overlap
+    ArrayList<TimeRange> mergedBusy = merge(busy); 
+
+    // Get the complement 
+    Collection<TimeRange> free = new ArrayList<>();
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: Same here about naming.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,101 @@
...
+      // Add time at end of day if possible
+      if (TimeRange.END_OF_DAY - mergedBusy.get(mergedBusy.size() - 1).end() >= request.getDuration()) {
+        free.add(TimeRange.fromStartEnd(mergedBusy.get(mergedBusy.size() - 1).end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    // Add the whole day if possible
+    else if (TimeRange.WHOLE_DAY.duration() >= request.getDuration() && request.getAttendees().size() > 0) {
+      free.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    }
+    return free;
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Consider moving some of the logic in the `if` conditions to methods, just to help make this more readable.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,101 @@
...
+    ArrayList<TimeRange> merged = new ArrayList<TimeRange>();
+
+    for (int i = 0; i < list.size(); i++) {
+      int lastMergedIndex = merged.size() - 1;
+      // Empty or there is no overlap, so add in the TimeRange
+      if (merged.size() == 0 || !merged.get(lastMergedIndex).overlaps(list.get(i))) {
+        merged.add(list.get(i));
+      }
+      // There is uncontained overlap, so replace last element
+      else if (!merged.get(lastMergedIndex).contains(list.get(i))) {
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: This `else if` should be in the same like as the `}` above.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +284,182 @@ public void notEnoughRoom() {
...
-}
 
+  @Test
+  public void severalOverlappingEvents() {
+    // Have one event with two people and two others overlapping . We should see three options.
+    //                 
+    // Events  :       |--A--|     |--B--|
+    //                 |--C--|        |--D--|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|         |3|
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:295]: Nice diagramming!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -270,5 +284,182 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_1030AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+	@Test
```
Comment (dslucas) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: Let's align `@Test` here and throughout this file.

**Review by dslucas — COMMENTED** [review]
Nice work with the algorithm and thinking through different test cases!

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java (hunk 1 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed - I moved this over to a new `complementOf` method

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Should this be on line line 95 here?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:295]: Thanks!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 7 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (dslucas) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Looks like line 78 to me, but in general if-else statements should read:

```
if (somecondtion) {
 // Do something
} else if (someothercondition) {
 // Do something else
} else {
 // Do something else else
}
```

**Review by dslucas — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (chrismohrii) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks, I fixed this for all if-else statements in the document

**Review by dborkan — APPROVED** [review]
👍 LGTM

**Pull request merged**